app/lib/three_scale/oauth2/service_discovery_client.rb

```diff
@@ -26,11 +26,60 @@ def uid
 
       def authenticate_options(request)
         {
-            # TODO: see service_discovery_helper.rb
-            # I need to construct this dynamically through a request
-            redirect_uri: 'http://master-account-admin.example.com.lvh.me:3000/auth/service-discovery/callback?self_domain=provider-admin.example.com.lvh.me'
+            redirect_uri: RedirectUri.call(self, request)
         }
       end
     end
+
+    # TODO: Refactor! It is the same as ThreeScale::OAuth2::RedhatCustomerPortalClient::RedirectUri
```

Comment:
Can you define a `ThreeScale::OAuth2::RedhatOAuthClient::RedirectUri` (or `ThreeScale::OAuth2::RedhatOAuthRedirectUri`) and make both inherit from it?

Reply:
I prefer letting that later. When I will do the tests I will see if it is necessary or not.